### PR TITLE
feat: Export as Delta/PDF/PlainText functionality.

### LINKED ISSUE #25: Support export to PDF
Refer documentation of [notes schema](https://github.com/SankethBK/diaryvault/blob/master/docs.md#3-notes)

We need to generate a PDF with note's `title`, `created_at`, `body` or `plainText`. We can use [pdf](Refer documentation of [notes schema](https://github.com/SankethBK/diaryvault/blob/master/docs.md#3-notes)) package to generate PDF.

Flutter quill also supports converting contents of rich_text editor to HTML. So there it might be possible to show rich text into PDF instead of plainText.

Challenge is to introduce images. We can retrieve all external asset paths of a note from `NoteDependencies` table.

### PR DESCRIPTION
**Pull request by mmjsmohit**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [X] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
This PR aims to add the export functionality to this app. This PR contains three options in the export menu - Export as Delta Object, Export as Plain Text and Export as a PDF.
## Featured Covered in this PR
- [x] Delta Object
- [X] Plain Text
- [X] PDF

## Related Tickets & Documents

- Closes #25 

##Screenshots, Recordings 
WhatsApp Image 2023-10-02 at 11 54 59 AM
image

## Tested Feature??

- [X] In Real Device.
- [X] In Emulator

### CODE REVIEW

**Comment by SankethBK** [comment]
Hey getting a blank popup here

[export_to_pdf.webm](https://github.com/SankethBK/diaryvault/assets/51091231/de1a7801-1336-4d4e-b9d7-ce875b445dce)

**Comment by mmjsmohit** [comment]
I don't think its a blank popup, I think the text and icon color are also black which I will fix in a commit shortly.

**Comment by mmjsmohit** [comment]
Hi! I have added dynamic styling to the share sheet. Now the text is properly visible as it is pulling the colour directly from the current theme.

The merge conflict in the `dropbox_sync_client.dart` is likely due to fact that I have fixed a typo.

File: lib/features/notes/core/converters/delta_to_pdf.dart [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+              fontSize = 16;
+            } else {
+              fontWeight = pw.FontWeight.bold;
+              fontSize = 12;
+            }
+            break;
+        }
+      });
+    }
+    return [fontWeight, fontSize];
```
Comment (micedreams) [inline lib/features/notes/core/converters/delta_to_pdf.dart:58]: please add a white line before the return statement.

File: lib/features/notes/core/converters/delta_to_pdf.dart [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+            break;
+          case "strike":
+            decoration = pw.TextDecoration.lineThrough;
+            break;
+          case "background":
+            boxDecoration = pw.BoxDecoration(color: PdfColor.fromHex(value));
+            break;
+        }
+      });
+    }
```
Comment (micedreams) [inline lib/features/notes/core/converters/delta_to_pdf.dart:92]: [same here](https://github.com/SankethBK/diaryvault/pull/45/files#r1349385722)

File: lib/features/notes/core/converters/delta_to_pdf.dart (hunk 1 above)
Comment (micedreams) [inline lib/features/notes/core/converters/delta_to_pdf.dart:58]:
```suggestion

    return [fontWeight, fontSize];
```
Please add white line before the return statement.

File: lib/features/notes/core/converters/delta_to_pdf.dart [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+      fontWeight = pw.FontWeight.bold;
+      fontSize = 24;
+    } else if (value == 2) {
+      fontWeight = pw.FontWeight.bold;
+      fontSize = 20;
+    } else {
+      fontWeight = pw.FontWeight.bold;
+      fontSize = 16;
+    }
+    return [fontWeight, fontSize];
```
Comment (micedreams) [inline lib/features/notes/core/converters/delta_to_pdf.dart:31]: Please add white line before the return statement.
```suggestion

    return [fontWeight, fontSize];
```

File: lib/features/notes/core/converters/delta_to_pdf.dart [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+/// [title] DeltaToPDF Class
+/// [Description] This is class with helper functions to convert flutter_quill delta object to pdf object
+/// [Written_by] mmjsmohit
+class DeltaToPDF {
+  /// [Description]  converts header text style
+  /// [Returns] fontWeight & fontSize
+  getPDFHeaderStyle(key, value) {
+    pw.FontWeight fontWeight = pw.FontWeight.normal;
+    double fontSize = 14;
```
Comment (micedreams) [inline lib/features/notes/core/converters/delta_to_pdf.dart:20]:
```suggestion
    double fontSize = 14;
    
```
Please add a white line after all declaration statements.

File: lib/features/notes/core/converters/delta_to_pdf.dart [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    }
+    return [fontWeight, fontSize];
+  }
+
+  /// [Description]  converts header attributed text style
+  /// [Returns] fontWeight & fontSize
+  getHeaderAttributedText(
+      Map<String, dynamic>? attribute, String text, bool hasAttribute) {
+    pw.FontWeight fontWeight = pw.FontWeight.normal;
+    double fontSize = 14;
```
Comment (micedreams) [inline lib/features/notes/core/converters/delta_to_pdf.dart:39]:
```suggestion
    double fontSize = 14;
    
```

File: lib/features/notes/core/converters/delta_to_pdf.dart [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+          case "strike":
+            decoration = pw.TextDecoration.lineThrough;
+            break;
+          case "background":
+            boxDecoration = pw.BoxDecoration(color: PdfColor.fromHex(value));
+            break;
+        }
+      });
+    }
+    return pw.Text(
```
Comment (micedreams) [inline lib/features/notes/core/converters/delta_to_pdf.dart:93]:
```suggestion

    return pw.Text(
```

File: lib/features/notes/core/converters/delta_to_pdf.dart [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+          fontFallback: [pw.Font.symbol()]),
+    );
+  }
+
+  /// [Description]  main wrapper to convert delta to pdf
+  /// [Returns] pdf widget object
+  deltaToPDF(List<fq.Operation> deltaList) {
+    List header = [null, null];
+    List texts = [];
+    List pdfColumnWidget = [];
```
Comment (micedreams) [inline lib/features/notes/core/converters/delta_to_pdf.dart:111]:
```suggestion
    List pdfColumnWidget = [];
    
```

File: lib/features/notes/core/converters/delta_to_pdf.dart [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+  }
+
+  /// [Description]  converts header attributed text
+  /// [Returns] all text with style
+  getAttributedText(Map<String, dynamic>? attribute, String text,
+      bool hasAttribute, pw.FontWeight fontWeight, double fontSize) {
+    PdfColor fontColor = PdfColor.fromHex("#000");
+    pw.FontStyle fontStyle = pw.FontStyle.normal;
+    pw.TextDecoration decoration = pw.TextDecoration.none;
+    pw.BoxDecoration boxDecoration = const pw.BoxDecoration();
```
Comment (micedreams) [inline lib/features/notes/core/converters/delta_to_pdf.dart:68]:
```suggestion
    pw.BoxDecoration boxDecoration = const pw.BoxDecoration();
    
```

File: lib/features/notes/core/converters/delta_to_pdf.dart [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+            element.attributes != null,
+            header[0] ?? pw.FontWeight.normal,
+            header[1] ?? 14));
+      }
+    }
+    pdfColumnWidget = pdfColumnWidget +
+        [
+          pw.Column(children: [...texts.reversed])
+        ];
+    return pw.Padding(
```
Comment (micedreams) [inline lib/features/notes/core/converters/delta_to_pdf.dart:139]:
```suggestion

    return pw.Padding(
```

File: lib/features/notes/presentation/widgets/note_export_button.dart [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+      bloc: notesBloc,
+      builder: (context, state) {
+        if (state.safe && state is! NoteSaveLoading) {
+          return Padding(
+            padding: const EdgeInsets.only(right: 13.0),
+            child: IconButton(
+              icon: const Icon(Icons.share),
+              onPressed: () {
+                _showBottomSheet(context, state);
+              },
```
Comment (micedreams) [inline lib/features/notes/presentation/widgets/note_export_button.dart:33]:
```suggestion
              onPressed: () => _showBottomSheet(context, state),
```

File: lib/features/notes/presentation/widgets/note_export_button.dart [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+  }
+}
+
+void _showBottomSheet(BuildContext context, NotesState state) {
+  showModalBottomSheet<void>(
+    context: context,
+    builder: (BuildContext context) {
+      final mainTextColor =
+          Theme.of(context).extension<PopupThemeExtensions>()!.mainTextColor;
+      return Container(
```
Comment (micedreams) [inline lib/features/notes/presentation/widgets/note_export_button.dart:66]:
```suggestion

      return Container(
```

File: packages/flutter_quill/flutter_quill_extensions/lib/embeds/builders.dart [hunk 12]
```diff
@@ -73,72 +73,72 @@ class ImageEmbedBuilder extends EmbedBuilder {
...
+                                controller
+                                  ..skipRequestKeyboard = true
+                                  ..formatText(
+                                      res.offset, 1, StyleAttribute(attr));
+                              },
+                              imageWidth: _imageSize?.width,
+                              imageHeight: _imageSize?.height,
+                              maxWidth: _screenSize.width,
+                              maxHeight: _screenSize.height);
+                        });
```
Comment (micedreams) [inline packages/flutter_quill/flutter_quill_extensions/lib/embeds/builders.dart:103]: Please turn this into a separate method.
Nesting too many times is not a good idea...

https://www.youtube.com/watch?v=CFRhGnuXG-4

File: packages/flutter_quill/flutter_quill_extensions/lib/embeds/builders.dart [hunk 13]
```diff
@@ -73,72 +73,72 @@ class ImageEmbedBuilder extends EmbedBuilder {
...
+                  onPressed: () {
+                    final offset =
+                        getEmbedNode(controller, controller.selection.start)
+                            .offset;
+                    controller.replaceText(
+                        offset, 1, '', TextSelection.collapsed(offset: offset));
+                    Navigator.pop(context);
+                  },
+                );
+                return Padding(
```
Comment (micedreams) [inline packages/flutter_quill/flutter_quill_extensions/lib/embeds/builders.dart:133]:
```suggestion

                return Padding(
```

**Review by micedreams — COMMENTED** [review]
Minor changes and this PR is probably good to go!!🎉

**Comment by mmjsmohit** [comment]
Thanks for the formatting tips. I think you can resolve the merge conflicts within the `dropbox_sync_client.dart` @SankethBK

**Comment by SankethBK** [comment]
@mmjsmohit slight change of plans, we added text to speech button in place of the share button that you currently have.
. 
I think the best place to have the share button here would be next to delete button in home page. So that users can export multiple notes as PDF at once. 

The main logic for PDF generation will still remain same as you implemented. It's just that we'll be moving `NoteShareButton` to home page app bar next to delete button. There will also be slight code changes related to generating the PDF for multiple notes instead of one. 

I will merge your PR to `export_as_pdf` branch which can be used as starting point for next changes, i will raise separate issue for this or refactor it myself. 

Thanks a lot for your contribution, I think this is the toughest functionality implemented in this app so far.

**Pull request merged**